### ISSUE TITLE
Receive mode seems to stall at 99% with large files

### ISSUE THREAD

**Issue by micahflee** [micahflee/onionshare#899]
While working on the 2.0 release, I ran into a bug in receive mode that is reproducible. It only seems to happen with large files (>100mb), but smaller files seem to work fine. When uploading a large file, the progress gets to 99%, and it appears like it's still transferring in both OnionShare and Tor Browser, but it never finishes.

**Comment by mig5** [comment]
I did some testing. Here is what I found. Uploading a 108MB file (created with `dd if=/dev/urandom`...) :

1) succeeds in native Firefox 60.5.0 ESR local mode, on Debian 9.7
2) fails in local mode on Tor Browser 8.0.5 (with Tor proxy disabled), on the same Debian 9.7 machine, which is also based on Firefox 60.5.0 ESR

2) fails in TB 8.0.5 when using Tor
3) fails in TB 8.0 when using Tor
4) succeeds in TB 7.5.6 (using Tor)
5) succeeds in TB 7.5 (using Tor)
6) succeeds with a torify curl POST request

Furthermore:

1) The version of Tor on the client side makes no difference
2) The version of Tor on OnionShare's connection side, makes no difference
3) v3 vs v2 onions make no difference

There is no version between TB 7.5.6 and 8.0, but 8.0 fails. We know from 8.0.5 that the underlying Firefox version works natively. Therefore, it suggests that between Tor 7.5.6 and 8.0, some change that is specific to TB's additions to Firefox, causes the upload to fail.

By 'fail', we mean the POST request to the /upload route just hangs forever in TBB 8.0+. 

In native Firefox, or with a curl request, or in TB 7.5.6, the expected 302 response fires ( `ReceiveModeRequest.upload_logic()` returns, and then the `ReceiveModeRequest.close()` function in OnionShare fires and returns, as expected. )

An example of it succeeding (native Firefox, local mode):

```
[Feb 11 2019 02:44:40 PM] ReceiveModeRequest.__init__
Feb 11, 02:44PM: Upload of total size 107.9 MiB is starting
[Feb 11 2019 02:44:40 PM] Web.check_slug_candidate: slug_candidate=silenced-apache
=> 307.4 KiB       file.txt[Feb 11 2019 02:44:40 PM] History.add: id: 0, item: <onionshare_gui.mode.history.ReceiveHistoryItem object at 0x7f6138ea1dc8>
=> 319.8 KiB       file.txt[Feb 11 2019 02:44:40 PM] ReceiveHistoryItemFile.__init__: filename: file.txt
=> 107.9 MiB       file.txt
[Feb 11 2019 02:47:09 PM] ReceiveModeWeb.define_routes: /upload, uploaded file.txt, saving to /home/user/OnionShare/2019-02-11/14.44.40/file.txt
Received: /home/user/OnionShare/2019-02-11/14.44.40/file.txt
[Feb 11 2019 02:47:10 PM] ReceiveModeRequest.close
127.0.0.1 - - [11/Feb/2019 14:47:10] "POST /silenced-apache/upload HTTP/1.1" 302 -
```

An example of it failing (Tor Browser 8.5.6, local mode):

```
[Feb 11 2019 02:47:19 PM] ReceiveModeRequest.__init__
Feb 11, 02:47PM: Upload of total size 107.9 MiB is starting
[Feb 11 2019 02:47:19 PM] Web.check_slug_candidate: slug_candidate=silenced-apache
=> 453.6 KiB       file.txt[Feb 11 2019 02:47:20 PM] History.add: id: 1, item: <onionshare_gui.mode.history.ReceiveHistoryItem object at 0x7f6138ea1828>
=> 458.1 KiB       file.txt[Feb 11 2019 02:47:20 PM] ReceiveHistoryItemFile.__init__: filename: file.txt
=> 107.9 MiB       file.txt
```

**Comment by mig5** [comment]
If I split my 107.9MB file into 2 halves, the 53.9MB file still hangs in Tor Browser 8.0.5.

If I split it into quarters, I succeed at uploading a 27.0MB file in Tor Browser 8.0.5.

**Comment by mig5** [comment]
We tracked the issue down to the NoScript addon.

Unchecking 'Sanitize cross-site suspicious requests' under the addon's advanced settings, or otherwise disabling the addon entirely, allows the file to be uploaded in Tor Browser 8.0.5.

Tor Browser 8.0 switched to the 'pure' WebExtension version of NoScript (version 10.1.91), so the problem probably began with that.

Unsure if there's anything else we can do to work properly with NoScript. There doesn't appear to be anything logged. Changing the site's 'trusted' status to something more trusted, makes no difference.

The next step would be to see if it is reproducible with NoScript on a non-OnionShare form (without javascript).

**Comment by micahflee** [comment]
The plot thickens...

I made this simple flask app that lets you upload arbitrary files: https://github.com/micahflee/noscript-upload-bug

I started it (`python3 app.py`) which started a server at http://127.0.0.1:5000/.

Then I open Tor Browser 8.0.5, opened Preferences, went to Network Proxy settings, and changed the proxy from "Manual proxy configuration" to "No proxy", to disable Tor in Tor Browser.

Then I loaded http://127.0.0.1:5000/, and successfully uploaded large files, including these two I'm using to test:

https://atom-installer.github.com/v1.34.0/atom-amd64.deb?s=1546643657&ext=.deb (85mb)
https://download.gimp.org/pub/gimp/v2.10/osx/gimp-2.10.8-x86_64-2.dmg (147mb)

I was expecting to trigger this NoScript bug, but I didn't. So then I started an OnionShare receive mode server with `--local-only`, and also successfully uploaded these two files from Tor Browser, without triggering the bug. -- But then, I closed OnionShare and tried again, and the second time, I reproduced it. Weird...

So I re-enabled Tor in Tor Browser, and started an OnionShare receive mode server, and uploaded `atom-amd64.deb` (85mb), and I confirm that I have hit the bug here too, but only while using Tor.

So, I made an onion service for my `noscript-upload-bug` flask app, and tried uploading `atom-amd64.deb` through the onion service -- and I hit the bug. By running `lsof -p [pid of flask app]` I could see that it has an open file that kept growing in size until it hit 89650515 bytes, then it just stalled. `atom-amd64.deb` is 89683126 bytes, which means it was failing to finish uploading the final 32611 bytes.

To recap:

- When going over Tor, in both OnionShare receive mode and `noscript-upload-bug`, I seem to always reproduce the bug with `atom-amd64.deb`
- When not going over Tor, the bug seems intermittent in OnionShare, but I haven't been able to reproduce it in `noscript-upload-bug`

**Comment by mig5** [comment]
Yeah, I didn't reproduce it in your PoC either. But in OnionShare it's always very consistent for me.

Then on a hunch I enabled Public Mode in OnionShare. After that, I reliably can upload.

I suspect something about the slug trips NoScript. The hyphen perhaps?

Since your PoC doesn't have a slug either...

**Comment by Baccount** [comment]
@mig5 Here is a quart from NoScripts FAQ that may relate to the slugs hyphen.

**Looks like the Anti-XSS feature causes problems with URLs containing some characters such as <, ' (single quote) or " (double quotes). What's happening?** 
A:   If you're following a link contained in an not trusted page and leading to a trusted page, this behaviour is expected by design. The reason is that those characters can be used to inject malicious code in the destination page, and since the source site is not trusted, "extreme" measures are taken by default.
https://noscript.net/faq#qa4_2

**Comment by mig5** [comment]
Nevermind, I reproduced the problem on a v3 onion with no slug. So it's either the whole domain, or 127.0.0.1 with a slug, but 127.0.0.1 by itself seems to be fine.

**Comment by micahflee** [comment]
I've opened an issue in NoScript https://github.com/hackademix/noscript/issues/64

**Comment by hackademix** [comment]
It seems this is not a NoScript-specific issue, but a bug in the webRequest.onBeforeRequest implementation happening when the "requestBody" option is enabled (I've tested with [a minimal extension](https://github.com/micahflee/onionshare/files/2855506/osbug.zip) which just registers an empty listener).
I'm further investigating to issue a meaningful report at https://bugzilla.mozilla.org (it might be related to https://bugzilla.mozilla.org/show_bug.cgi?id=1506562 ), but in the meanwhile I'd suggest to switch to an Ajax-based upload, since it would bypass the XSS filter listener which is filtered by "main_frame", "sub_frame" and "object".

**Comment by micahflee** [comment]
Thanks so much for looking into this @hackademix, and for investigating enough to open an upstream issue in Firefox.

I'm curious, does SecureDrop have this same problem? It seems like they should. Their upload page also doesn't use Ajax, and is very similar to ours, and their web server is also python and flask. I'll ask the SecureDrop people.

I want OnionShare to work for Tor Browser users that set their security sliders to Safest, which disables Javascript and would prevent an Ajax uploader from working. But I do think that's a minority of Tor Browser users since the default is Standard. I think a reasonable solution would be to implement an Ajax uploader and use that in most cases, but also have a `<noscript>` tag for people with the Safest setting. It could allow uploading without Ajax, but include a brief explanation of the bug -- basically saying that, due to a bug, if they're planning on uploading greater than 50mb, to set their security slider to Safer to enable Javascript and use the Ajax uploader.

Another benefit to making an Ajax uploader is the person sending the files can actually get a progress bar on their end too, like when you upload a file via https://send.firefox.com/, which will have a better UX on that side. If we're going to do a bunch of work on the HTML side of things, I think we might as well make those pages response (#886) so they look good in mobile too, while we're at it.

**Comment by micahflee** [comment]
It turns out that SecureDrop has this exact same issue: https://github.com/freedomofpress/securedrop/issues/4078

**Comment by eloquence** [comment]
@micahflee FYI, I was able to immediately reproduce the issue with your [noscript-upload-bug Flask app](https://github.com/micahflee/noscript-upload-bug) on Tor 8.0.6 against 127.0.0.1.

Note the high number of successful attempts, so it is quite sporadic. But it appears to be in no way dependent on onion services.

Contrary to https://bugzilla.mozilla.org/show_bug.cgi?id=1506562 I see no error message on the console.

## Attempt 1

Stalled out:

|  | Byte size           | Δ Difference |
| ------------- |-------------:| -------------:| 
|Uploaded file | 283,519,269 bytes| |
| Size at stall point per `lsof` | 283,504,383 bytes|14,886 bytes
| Size reported by `app.py` | 283,507,891 bytes |11,378 bytes

## Attempt 2

Completed successfully.

## Atttempt 3

Stalled out. Exact same results as first try:

|  | Byte size           | Δ Difference |
| ------------- |-------------:| -------------:| 
|Uploaded file | 283,519,269 bytes| |
| Size at stall point per `lsof` | 283,504,383 bytes|14,886 bytes
| Size reported by `app.py` | 283,507,891 bytes |11,378 bytes

## Attempts 4-11

Completed successfully (across unsystematic restarts, file deletions).

## Attempt 12

Stalled out. Exact same results as first try:

|  | Byte size           | Δ Difference |
| ------------- |-------------:| -------------:| 
|Uploaded file | 283,519,269 bytes| |
| Size at stall point per `lsof` | 283,504,383 bytes|14,886 bytes
| Size reported by `app.py` | 283,507,891 bytes |11,378 bytes

**Comment by eloquence** [comment]
I can confirm that @hackademix' [test extension](https://github.com/micahflee/onionshare/files/2855506/osbug.zip) causes uploads to intermittently stall out against Micah's sample upload form exactly the same manner. 

I've been unable to reproduce the issue if the `requestBody` option is omitted in the extension code. https://bugzilla.mozilla.org/show_bug.cgi?id=1506562 does describe similarly problematic behavior, but I'm not seeing any of the errors described, and  use of XMLHttpRequest is not necessary to trigger the problem. I think a new bug report is warranted, for clarity alone.

I'll ping @hackademix by email as well, and if he doesn't beat me to it, I'll file a more detailed report in Bugzilla. Let's also make an effort to reach folks at Mozilla so they're aware of the scale of impact this bug is having across the web.

**Comment by eloquence** [comment]
I've filed https://bugzilla.mozilla.org/show_bug.cgi?id=1532530 and cross-linked it from the previous issue.

**Comment by eloquence** [comment]
@micahflee @mig5 There's a NoScript RC up which may resolve this with the next Tor release, check out https://trac.torproject.org/projects/tor/ticket/29733#comment:14 in case you have time to do your own test(s). :)

**Comment by mig5** [comment]
Thank you for keeping us in the loop @eloquence ! I'm testing it now following your steps in that ticket. Will report back soon after a couple of iterations.

**Comment by mig5** [comment]
@eloquence @micahflee I replied in the Trac ticket but yes, it seems to fix the issue for OnionShare too! Thanks again for the coordination you've done on this and for posting your steps to test the new RCs.

**Comment by mig5** [comment]
Looks like the NoScript fixes are now in the stable release 10.2.3.

Unfortunately, it looks like Tor Browser 8.0.7 shipped a couple hours earlier with 10.2.1. So we'll have to wait for yet another TB release before we can take down our NoScript 'warning' content...

**Comment by eloquence** [comment]
ICYMI, it made its way into 8.0.8, which is also part of the latest Tails release:
https://blog.torproject.org/new-release-tor-browser-808

We're planning to remove the instructions with an upcoming point release of SecureDrop, pending QA. As you probably saw there's upstream resolution as well in https://bugzilla.mozilla.org/show_bug.cgi?id=1532530 , but I don't know how long it will take for that to land in Tor.
